include/chainbase/undo_index.hpp

```diff
@@ -48,16 +48,25 @@ namespace chainbase {
    };
 
    template<class Tag>
-   struct offset_node_base {
+   struct __attribute__((packed, aligned(4))) offset_node_base {
       offset_node_base() = default;
       offset_node_base(const offset_node_base&) {}
       constexpr offset_node_base& operator=(const offset_node_base&) { return *this; }
-      std::ptrdiff_t _parent;
-      std::ptrdiff_t _left;
-      std::ptrdiff_t _right;
-      int _color;
+      int64_t _parent:42;
+      int64_t _left  :42;
+      int64_t _right :42;
+      int64_t _color :2;
    };
 
+   // --------------------------------------------------------------------------------------
+   // Because the pointers are always aligned to an 4 byte boundary
+   // (so the 2 least significant bits are always 0), we store pointer offsets
+   // shifted by two bits, extending our maximum memory support from 2^41 = 2TB
+   // to 2^43 = 8TB.
+   // A concern could be that `1` is a special value meaning `nullptr`. However we could not
+   // get an offset of 4, since `sizeof(offset_node_base) == 16`, so we could not have
+   // difference between two different `node_ptr` be less than 16.
+   // --------------------------------------------------------------------------------------
```

Comment:
Did you run any performance comparisons with these changes?

Reply:
No, I didn't. Do you know of a good way to make a performance comparison? Maybe it would require to write a small benchmarking test.

Reply:
You could write one I suppose, but you could also just do a replay comparison of some range of blocks.

Reply:
I think I'll write a small test.